alf/algorithms/sarsa_algorithm.py

```diff
@@ -13,41 +13,136 @@
 # limitations under the License.
 """SARSA Algorithm."""
 
+from absl import logging
+import copy
 import gin
-import tensorflow as tf
-from tf_agents.networks.network import Network, DistributionNetwork
-from tf_agents.agents.ddpg.critic_network import CriticNetwork
-from tf_agents.agents.ddpg.critic_rnn_network import CriticRnnNetwork
-from tf_agents.networks.actor_distribution_network import ActorDistributionNetwork
-from tf_agents.networks.actor_distribution_rnn_network import ActorDistributionRnnNetwork
-from tf_agents.trajectories.time_step import StepType
-from tf_agents.utils import common as tfa_common
-
-from alf.algorithms.ddpg_algorithm import create_ou_process
+import numpy as np
+import torch
+import torch.nn as nn
+
+import alf
+from alf.algorithms.algorithm import Algorithm
+from alf.algorithms.one_step_loss import OneStepTDLoss
+from alf.algorithms.rl_algorithm import RLAlgorithm
 from alf.algorithms.on_policy_algorithm import OnPolicyAlgorithm
-from alf.data_structures import ActionTimeStep, LossInfo, PolicyStep, TrainingInfo
-from alf.data_structures import namedtuple
-from alf.utils import common, dist_utils, losses
+from alf.data_structures import (AlgStep, Experience, experience_to_time_step,
+                                 LossInfo, namedtuple, StepType, TimeStep,
+                                 TrainingInfo)
+from alf.networks import Network, DistributionNetwork
+from alf.utils import common, dist_utils, losses, math_ops, tensor_utils
+from alf.utils.averager import AdaptiveAverager
 from alf.utils.summary_utils import safe_mean_hist_summary
 
 SarsaState = namedtuple(
     'SarsaState', [
-        'prev_observation', 'prev_step_type', 'actor', 'target_actor',
-        'critic', 'target_critic'
+        'prev_observation', 'prev_step_type', 'actor', 'critics',
+        'target_critics'
     ],
     default_value=())
 SarsaInfo = namedtuple(
-    'SarsaInfo', ['action_distribution', 'actor_loss', 'critic', 'returns'])
-SarsaLossInfo = namedtuple(
-    'SarsaLossInfo', ['actor', 'critic'], default_value=())
+    'SarsaInfo', [
+        'action_distribution', 'actor_loss', 'alpha_loss', 'critics',
+        'target_critics', 'entropy_reward'
+    ],
+    default_value=())
+SarsaLossInfo = namedtuple('SarsaLossInfo', ['actor', 'critic', 'alpha'])
+
+nest_map = alf.nest.map_structure
+
+
+class FastCriticBias(nn.Module):
+    """
+    The motivation of this is to prevent the critic value becomes greater and
+    greater as training goes on because of the max operation.
```

Comment:
SARSA doesn't have any max operation? because the target critic uses the actual action

Reply:
action is trained to maximize the q value. It is an implicit max operation. And for off-policy training, the action used to calculate the target_critic is not from experience, it is calculated every time.